Consolidate `Target` mappings

raynelfss opened the pull request:
<!--
⚠️  If you do not respect this template, your pull request will be closed.
⚠️  Your pull request title should be short detailed and understandable for all.
⚠️  Also, please add a release note file using reno if the change needs to be documented in the release notes.
⚠️  If your pull request fixes an open issue, please link to the issue. Use "Fixes #XXXX" if this PR *fully* closes the issue XXXX.  
☢️  If you used an AI tool to code this PR, add "AI tool used: <Name and version of the tool>". For example, "AI tool used: Microsoft Copilot Chat with GPT-5". Failing to disclose the use of AI tools may result in the PR being closed without further review.  


- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary
The following commits remove the redundant mappings `gate_name_map` and `angle_bounds` in favor of one mapping with a struct that contains all of that data. This results in a much lighter `Target` going from 504 (0x1F8) bytes to 288 bytes.

The new struct called `TargetProperties` is currently internal-only. And contains three fields.
- `properties` containing the mapping of the qargs and instruction properties (originally the `gate_map`).
- `instruction` containing the `TargetOperation` (originally the `gate_name_map`).
- `angle_bounds`: Optionally containing the angle bounds for the ``Target`` (originally `angle_bounds`.)

This is the first effort to more efficiently use the ``Target`` in Rust with upcoming efforts focusing on its behavior on Python.


### Details and comments

raynelfss committed:
Initial: Clean up target mappings
The following commits remove the redundant mappings `gate_name_map` and `angle_bounds` in favor of one mapping with a struct that contains all of that data. This results in a much lighter `Target` going from 504 (0x1F8) bytes to 288 bytes.

The new struct called `TargetProperties` is currently internal-only. And contains three fields.
- `properties` containing the mapping of the qargs and instruction properties (originally the `gate_map`).
- `instruction` containing the `TargetOperation` (originally the `gate_name_map`).
- `angle_bounds`: Optionally containing the angle bounds for the ``Target`` (originally `angle_bounds`.)

This is the first effort to more efficiently use the ``Target`` in Rust with upcoming efforts focusing on its behavior on Python.

raynelfss committed:
Fix: Remove clippy unused

raynelfss committed:
Merge remote-tracking branch 'upstream/main' into target-lighter

Cryoris commented on crates/transpiler/src/target/mod.rs:
```diff
@@ -181,6 +185,17 @@ impl Clone for NormalOperation {
 
+/// A collection of all the properties of an instruction in the [Target]
+#[derive(Debug, Clone)]
+pub(crate) struct TargetProperties {
```
We can make this private since it's only being used within this file, right?

raynelfss replied:
Fixed in 6e7aaa83e3cb878d23f34d5df832af02507f11cb

mtreinish commented on crates/transpiler/src/target/mod.rs:
```diff
@@ -219,12 +234,10 @@ pub struct Target {
-    qarg_gate_map: IndexMap<Qargs, HashSet<String>, RandomState>,
-    angle_bounds: HashMap<String, AngleBound>,
+    gate_map: IndexMap<String, TargetProperties, RandomState>,
+    global_operations: HashMap<u32, HashSet<String>>,
+    qarg_gate_map: HashMap<Qargs, HashSet<String>>,
```
I assume these were an indexmap previously for deterministic iteration order which would have been important for Python backwards compatibility (these were originally python dicts I think). Are these not exposed directly in the public api for the target, or were there no methods to iterate them and that's why it's safe to change?

raynelfss replied:
Yes, in most cases I noticed the iteration was either internal only for classification, or the iteration was exposed to `Python` as a `set`, which in `Python` is unordered. For example, the `Target::qargs` method which is the only one exposed to python that iterates through one of these maps `qarg_gate_map`, wraps everything inside of a set, while the rust version returns an iterator.

mtreinish commented on crates/transpiler/src/target/mod.rs:
```diff
@@ -219,12 +234,10 @@ pub struct Target {
     pub qubit_properties: Option<Vec<QubitProperties>>,
     #[pyo3(get, set)]
     pub concurrent_measurements: Option<Vec<Vec<PhysicalQubit>>>,
-    gate_map: GateMap,
```
I'm very happy to see this type alias go away, I always found the type aliases got in the way, and seeing that this is an Indexmap with the types is much easier to reason about.

mtreinish commented on crates/transpiler/src/target/mod.rs:
```diff
@@ -344,24 +356,19 @@ impl Target {
         instruction: TargetOperation,
-        name: String,
+        name: &str,
         properties: Option<PropsMap>,
```
Can we drop the use of props map here too and just write out the type?

raynelfss replied:
I replaced the type by a "simplified" version that omits the hasher in 6e7aaa83e3cb878d23f34d5df832af02507f11cb

mtreinish commented on crates/transpiler/src/target/mod.rs:
```diff
@@ -688,8 +689,8 @@ impl Target {
     // TODO: Add flag for custom tests
     /// Private method for development purposes only
```
Is this still true? If this method isn't used we should just delete it.

raynelfss replied:
It had an associated test with it, which I removed in 6e7aaa83e3cb878d23f34d5df832af02507f11cb

mtreinish commented on crates/transpiler/src/target/mod.rs:
```diff
@@ -833,27 +843,40 @@ impl Target {
-        self.gate_map = state.get_item("gate_map")?.unwrap().extract::<GateMap>()?;
-        self._gate_name_map = state
-            .get_item("gate_name_map")?
+        let mut gate_map = IndexMap::default();
```
Don't you know the size from the dict? So I'd expect you be able to use `with_capacity` and  pre-allocate here.

raynelfss replied:
Added in 6e7aaa83e3cb878d23f34d5df832af02507f11cb

mtreinish commented on crates/transpiler/src/target/mod.rs:
```diff
@@ -978,57 +1006,65 @@ impl Target {
-                props_map = IndexMap::from_iter([(Qargs::Global, None)]);
-            }
+        let properties = match instruction {
+            TargetOperation::Variadic(_) => IndexMap::from_iter([(Qargs::Global, None)]),
```
I think this would work and be a bit more concise: 

```suggestion
            TargetOperation::Variadic(_) => indexmap!{Qargs::Global => None},
```

see: https://docs.rs/indexmap/latest/indexmap/macro.indexmap.html

raynelfss replied:
I tried doing this but it doesn't seem to work too well currently because of two reasons:
- `indexmap! {}` uses the default hasher and here we require usage of `ahash::RandomState`.
- `indexmap_with_default! {}` wraps the custom hasher inside of a `BuildHasherDefault` which doesn't evaluate to the same types here and results in the compiler yelling about not evaluating the match arm correctly.
    ```expected IndexMap<{unknown}, {unknown}, BuildHasherDefault<RandomState>>, found IndexMap<Qargs, Option<InstructionProperties>, RandomState>```

mtreinish commented on crates/transpiler/src/target/mod.rs:
```diff
@@ -978,57 +1006,65 @@ impl Target {
                         if qarg_slice.len() != instruction.num_qubits() as usize {
                             return Err(TargetError::QargsMismatch {
-                                instruction: name,
+                                instruction: name.to_string(),
```
I know I'm the one always going on about using &str in rust interfaces. But I wonder if in this case passing an owned string to this inner method might be a better choice. Then we could avoid this copy. If you have the python method take an owned string, it shouldn't need a clone() anymore since everything else in the python `add_instruction()` should be by reference after this change. So if you took an owned string in from python and moved it here we'd avoid a clone() or to_string() I think. For a rust/c public/outward facing interface we can still work with &str if that's more ergonomic and do the copy before calling the inner method.

raynelfss replied:
I think it would make sense since, after all we are keeping the string inside the mapping, so I agree with this assessment. That said, it might be too late to include that in this PR as it would change the input of one of the most important methods in this struct. Perhaps we might do this for a future PR I have planned.

mtreinish replied:
I'm not sure I follow, this PR hasn't merged yet how is it too late? I think most of what I was proposing was just undoing changes that this is making and removing clones/to_string calls that aren't needed with the new internal data structure.

If you're thinking about the 2.40rc1 scheduled for today I don't think we should rush to include this in the release and should just do it for 2.5.0. Even if it were ready a change to a core data model like this is not something you want to merge the day of a release, it's too risky since there is no margin to catch mistakes before release.

raynelfss replied:
Oh, I'm sorry. This PR has been open for a bit and I had forgotten about introducing the changes here. Yes I can revert it now.

mtreinish commented on crates/transpiler/src/target/mod.rs:
```diff
@@ -1501,18 +1538,14 @@ impl Target {
 
-    /// Retrieve all the gate names in the Target
-    // TODO: Remove once `Target` is being consumed.
-    #[allow(dead_code)]
```
Something is using it now?

raynelfss replied:
Nothing is using it currently but I wanted to make sure that if it weren't it wouldn't be supressed by this flag.

mtreinish reviewed (commented):
Overall I really like the structural change here, it reduces the weight of the target more than you pointed out in the PR summary message. That only looks mechanically at the data in the struct itself, not the allocations being pointed too. I expect for a typical target this will shrink the memory footprint more substantially. You've eliminated 2 string copies per instruction in the target. Did you run any benchmarks with this to see if it improved things at all? Also did we have an issue about reducing the string copies in the target at one point? If we did we should mark this as closing it.

I just had a few inline comments and questions.

Cryoris reviewed (commented):
Matt superseded the remaining comments I had, so only 1 is left 🤷🏻‍♂️

raynelfss committed:
Fix: Address review comments
- Replace ambiguous type declarations with more explicit types.
    - `PropsMap` to `TargetIndexMap<K,V>` == `IndexMap<K,V,RandomState>`.
- Make `TargetProperties` private.
- Remove `_raw_operation_from_name`.
- Use `IndexMap::with_capacity_and_hasher` during deserialization.
- Revert overusage of `&str`.

raynelfss committed:
Refactor: Remove type alias.
- In cases where the alias of `IndexMap<String, TargetProperties, ahash::RandomState>` looked too long, I matched it to the `Index` trait whose return `Output` value is already of said type.

raynelfss committed:
Merge branch 'main' into target-lighter

raynelfss committed:
Merge remote-tracking branch 'upstream/main' into target-lighter

mtreinish reviewed (dismissed):
This LGTM now, thanks for the updates. Once you fix the lint issue we can merge this

raynelfss committed:
Fix: unused Barrier import in test_target.py
